src/data-connect/data-connect-api-client-internal.ts

```diff
@@ -138,28 +145,138 @@ export class DataConnectApiClient {
+    endpoint: string,
+    options: GraphqlOptions<Variables>
+  ): Promise<ExecuteGraphqlResponse<GraphqlResponse>> {
+    if (
+      typeof options.operationName === 'undefined' ||
+      !validator.isNonEmptyString(options.operationName)
+    ) {
+      throw new FirebaseDataConnectError(
+        DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT,
+        '`options.operationName` must be a non-empty string.'
+      );
+    }
+    if (
+      typeof options.impersonate === 'undefined' || 
+      !validator.isNonNullObject(options?.impersonate)
+    ) {
+      throw new FirebaseDataConnectError(
+        DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT,
+        '`options.impersonate` must be a non-null object.'
+      );
+    }
+
+    if (this.connectorConfig.connector === undefined || this.connectorConfig.connector === '') {
+      throw new FirebaseDataConnectError(
+        DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT,
+        `The 'connectorConfig.connector' field used to instantiate your Data Connect
+        instance must be a non-empty string (the connectorId) when calling executeQuery or executeMutation.`);
+    }
+
+    const data = {
+      ...(options.variables && { variables: options?.variables }),
+      operationName: options.operationName,
+      extensions: { impersonate: options.impersonate },
+    };
+    const url = await this.getUrl(
+      API_VERSION,
+      this.connectorConfig.location,
+      this.connectorConfig.serviceId,
+      endpoint,
+      this.connectorConfig.connector,
+    );
+    return this.makeGqlRequest<GraphqlResponse>(url, data)
+      .then((resp) => {
+        return resp;
+      })
+      .catch((err) => {
+        throw this.toFirebaseError(err);
       });
   }
 
+  /**
+   * Constructs the URL for a Data Connect backend request.
+   * 
+   * If no connectorId is provided, will direct the request to an endpoint under services:
+   * .../services/{serviceId}:endpoint
+   *
+   * If connectorId is provided, will direct the request to an endpoint under connectors:
+   * .../services/{serviceId}/connectors/{connectorId}:endpoint
+   * 
+   * @param version - The API version.
+   * @param locationId - The location of the Data Connect service.
+   * @param serviceId - The ID of the Data Connect service.
+   * @param endpointId - The endpoint to call.
+   * @param connectorId - The ID of the connector, if applicable.
+   * @returns A promise that fulfills with the constructed URL.
+   */
+  private async getUrl(
+    version: string,
+    locationId: string,
+    serviceId: string,
+    endpointId: string,
+    connectorId?: string,
+  ): Promise<string> {
+    const projectId = await this.getProjectId();
+    const urlParams = {
+      version,
+      projectId,
+      locationId,
+      serviceId,
+      endpointId,
+      connectorId
+    };
+    let urlFormat: string;
+    if (useEmulator()) {
+      (urlParams as any).host = emulatorHost();
```

Comment:
Can we make this more strongly typed? I'd rather we add an empty host field in `urlParams` than cast to `any` and lose type-safety